<br><br><br><br><br><br><br><br>

# Lesson 1: Basic Queries and the American Community Survey (ACS)

We will start by querying the median household income of Washington
State residents in 2022 and move on from there.

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Query Basics

Every query needs at least four pieces of information.

1. The dataset. We will start wih the American Community Survey.
2. The vintage, or year, of the data we want to query. We will use 2022.
3. The variables we want to query. We'll start with median income.
4. The geography or geographies we want to query data for. Since we are in Washington State, we'll
   do our first query for Washington State.

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

### The American Community Survey (ACS)

- The U.S. Census data is organized into many different data sets. One of the most
commonly used is the [American Community Survey (ACS)](https://www.census.gov/programs-surveys/acs/). 

- According to the U.S. Census Bureau, "it is the premier source for detailed population and housing information about our nation."

The ACS survey data is aggregates over 1, 3, and 5 year periods and published every year.

Each data set has a corresponding identifier. `censusdis` has symbolic names for all of them, and is
updated daily to accomodate new data sets as they are published.

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Imports and Configuration

The first thing we will import in almost all cases is `censusdis.data`. The standard
convention is to import it as `ced`. This module contains the key APIs for downloading
data from the U.S. Census API.

In [1]:
import censusdis.data as ced


Next, we will import the identifier for the ACS 5-year data set, which aggregates survey results over
the trailing five years.

In [2]:
from censusdis.datasets import ACS5

Finally, we will import the symbolic names for all the states. For example, `states.WA` for Washington State.

In [3]:
import censusdis.states as states

<br>
The ACS consists of thousands of variables. The one we want is median household income.
It is called `B19013_001E`. In a later lesson, we will talk about how we found this variable.

In [4]:
# See https://api.census.gov/data/2020/acs/acs5/variables/B19013_001E.html
MEDIAN_HOUSEHOLD_INCOME_VARIABLE = "B19013_001E"

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Our first query

In [5]:
df_wa_income = ced.download(
    
    # Data set: American Community Survey 5-Year
    dataset=ACS5,
    
    # Vintage: 2022
    vintage=2022, 
    
    # Variable: median household income
    download_variables=MEDIAN_HOUSEHOLD_INCOME_VARIABLE,
    
    # Geography: Washington State
    state=states.WA
)

In [6]:
df_wa_income

STATE  B19013_001E
0    53        90325

### Interpreting the Results

We asked for data from a single geography, Washington State, so the return value is a data frame with only one row.

The data frame has two columns:

- `STATE`, the [FIPS code](https://transition.fcc.gov/oet/info/maps/census/fips/fips.txt) for Washington State. This is also the value of `censusdis.states.WA`.
- `B19013_001E`, the value of variable we asked for, which is an estimate of median household income*.

So, according to this data, the median household income in Washington State in 2022 was estimated to be $90,325. 

In [7]:
# Verify the FIPS code for WA.
states.WA

'53'

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Extending our Query to More than One State

We don't have to limit our queries to just a single geography.

We can query several states at onece.

In [8]:
df_west_coast_income = ced.download(
    dataset=ACS5,
    vintage=2022,
    download_variables=MEDIAN_HOUSEHOLD_INCOME_VARIABLE, 
    
    # Three states this time.
    state=[states.WA, states.OR, states.CA]  
)

df_west_coast_income

STATE  B19013_001E
0    06        91905
1    41        76632
2    53        90325

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Extension: More than One Variable

We can also query several variables at once.

We'll add the variable `NAME` to our query so it is easier for those of us who don't dream in FIPS codes to know what state each row represents.

In [9]:
df_west_coast_income = ced.download(
    dataset='acs/acs5',
    vintage=2020,
    
    # Name and median household income
    download_variables=['NAME', MEDIAN_HOUSEHOLD_INCOME_VARIABLE],  
    
    state=[states.WA, states.OR, states.CA]
)

df_west_coast_income

STATE        NAME  B19013_001E
0    06  California        78672
1    41      Oregon        65667
2    53  Washington        77006

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Extension: All the states

We can use `'*'` as a wildcard to get all known geographies of a given
type.

In [10]:
df_all_states = ced.download(
    dataset='acs/acs5',
    vintage=2022,
    download_variables=['NAME', MEDIAN_HOUSEHOLD_INCOME_VARIABLE], 
    
    # Wild card for all the states
    state='*'
)

df_all_states

STATE                  NAME  B19013_001E
0     01               Alabama        59609
1     02                Alaska        86370
2     04               Arizona        72581
3     05              Arkansas        56335
4     06            California        91905
5     08              Colorado        87598
6     09           Connecticut        90213
7     10              Delaware        79325
8     11  District of Columbia       101722
9     12               Florida        67917
10    13               Georgia        71355
11    15                Hawaii        94814
12    16                 Idaho        70214
13    17              Illinois        78433
14    18               Indiana        67173
15    19                  Iowa        70571
16    20                Kansas        69747
17    21              Kentucky        60183
18    22             Louisiana        57852
19    23                 Maine        68251
20    24              Maryland        98461
21    25         Massachusetts        96505
22    26              Michigan        68505
23    27             Minnesota        84313
24    28           Mississippi        52985
25    29              Missouri        65920
26    30               Montana        66341
27    31              Nebraska        71722
28    32                Nevada        71646
29    33         New Hampshire        90845
30    34            New Jersey        97126
31    35            New Mexico        58722
32    36              New York        81386
33    37        North Carolina        66186
34    38          North Dakota        73959
35    39                  Ohio        66990
36    40              Oklahoma        61364
37    41                Oregon        76632
38    42          Pennsylvania        73170
39    44          Rhode Island        81370
40    45        South Carolina        63623
41    46          South Dakota        69457
42    47             Tennessee        64035
43    48                 Texas        73035
44    49                  Utah        86833
45    50               Vermont        74014
46    51              Virginia        87249
47    53            Washington        90325
48    54         West Virginia        55217
49    55             Wisconsin        72458
50    56               Wyoming        72495
51    72           Puerto Rico        24002

## The results of our queries are ordinary pandas [DataFrames](https://pandas.pydata.org/docs/reference/frame.html)

So we can do whatever analysis we might normally do with a data frame.
For example, what are the top ten states by median income?

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

In [11]:
df_all_states.nlargest(10, MEDIAN_HOUSEHOLD_INCOME_VARIABLE)

STATE                  NAME  B19013_001E
8     11  District of Columbia       101722
20    24              Maryland        98461
30    34            New Jersey        97126
21    25         Massachusetts        96505
11    15                Hawaii        94814
4     06            California        91905
29    33         New Hampshire        90845
47    53            Washington        90325
6     09           Connecticut        90213
5     08              Colorado        87598

Or the bottom ten?

In [12]:
df_all_states.nsmallest(10, MEDIAN_HOUSEHOLD_INCOME_VARIABLE)

STATE            NAME  B19013_001E
51    72     Puerto Rico        24002
24    28     Mississippi        52985
48    54   West Virginia        55217
3     05        Arkansas        56335
18    22       Louisiana        57852
31    35      New Mexico        58722
0     01         Alabama        59609
17    21        Kentucky        60183
36    40        Oklahoma        61364
40    45  South Carolina        63623

<br><br><br><br><br><br><br><br>
<br><br><br><br><br><br><br><br>

## Exercise 1

Now it's time to get hands-on and apply what you have learned.
[Exercise 1](../Exercises/Exercise%201.ipynb) is a notebook
where you can make census queries similar to the ones we just 
made.